colcon_core/dependency_descriptor.py

```diff
@@ -34,3 +36,11 @@ def name(self):
         :rtype: str
         """
         return str(self)
+
+    def __deepcopy__(self, memodict):  # noqa: D105
+        # surprisingly this is significantly faster than the default
+        if not self.metadata:
+            # explicitly skipping the deep copy of an empty dict is also faster
+            return DependencyDescriptor(self.name)
+        return DependencyDescriptor(
+            self.name, metadata=copy.deepcopy(self.metadata))
```

Comment:
According to the [`copy.deepcopy()` docs](https://docs.python.org/3/library/copy.html), you should pass `memodict` to the `copy.deepcopy()` call from within `__deepcopy__()`. The docs outline why this is necessary to handle all cases.

Could the lack of a `memodict` within the metadata's deep copy be contributing to the perceived performance gains?